### ISSUE TITLE
No login to geocaching.com with Android 4.0.3

### ISSUE THREAD

**Issue by murmeltier4** [cgeo/cgeo#7078]
Since a few days it is not longer possible to login to geocaching.com.
Error message: "Anmeldung fehlgeschlagen: Verbindung zu Geocaching.com konnte nicht hergestellt werden (Server oder Verbindung inaktiv?)"
Does not work with the newest 2018.06.26 as well as the older 2018.02.05
Seems to be a problem with the Android 4.0.3 version (https-Login?) on the HTC one V; because in Android 8.0 it works well with the same account.
Unfortunatly one day before starting to holidays... Should buy a newer smartphone...

**Comment by Lineflyer** [comment]
Does it work with your browser on that phone to login?
Are date/time set correctly?

**Comment by murmeltier4** [comment]
With the default phone browser not (https-problem) - website is not shown. With google chrome login is possible. Date/time are OK.

**Comment by Lineflyer** [comment]
Did you try to reset your login as described here: http://faq.cgeo.org/#login-failed

**Comment by murmeltier4** [comment]
Bitte melde dich direkt auf der Geocaching-Website (mit deinem Smartphone-Browser oder PC) an und prüfe, dass dein Zugang dort nicht deaktiviert ist und deine Mail-Adresse auf der Webseite bestätigt ist.
WORKS

Bitte prüfe, dass dein Telefon das richtige Datum und die richtige Uhrzeit hat. Wenn das Datum nicht korrekt ist, kann es zu Problemen kommen eine sichere Verbindung zur Webseite aufzubauen.
DATE/TIME OK

Manchmal fügen Netzwerk-Betreiber Nachrichten in den HTML-Verkehr ein, die c:geo davon abhalten sich in einem Mobilfunknetz erfolgreich einzuloggen. Ein klares Indiz für dieses Problem ist, dass c:geo nur mit deiner WLAN-Verbindung funktioniert, aber nicht im Mobilfunk-Netz. 
NO LOGIN WITH WLAN OR MOBILE DATA
Gehe in diesem Fall bitte in das c:geo Menü → Einstellungen → Dienste und aktiviere “Als Android-Browser identifizieren”. Danach musst du evtl. dein Telefon neu starten, um sicherzustellen, dass die Einstellung aktiv ist.
I DID SO, BUT SAME ERROR-MESSAGE

Es gab Berichte, dass in seltenen Fällen die Anmeldung trotz korrekter Nutzerdaten fehlschlägt und auch “Als Android-Browser identifizieren” nicht hilft. In diesem Fall empfehlen wir deine Logindaten zurückzusetzen, indem du lange auf das Feld “c:geo autorisieren” in den Plattform-Einstellungen unter Menü → Einstellungen → Dienste klickst. Danach solltest du dein Gerät neu starten und die Anmeldedaten wie oben beschrieben neu eingeben.
I DID SO, RESTART, INPUT NEW LOGIN DATA, BUT SAME ERROR-MESSAGE

Als letzte Möglichkeit kannst du c:geo auf Werkseinstellungen zurückzusetzen, wie es hier beschrieben ist.
SAME ERROR

I STILL THINK ITS A PROBLEM WITH THE SSL-CERTIVICATE USED BY C:GEO TO LOGIN. BECAUSE DEFAULT BROWSER DOES NOT WORK TO CONNECT TO GEOCACHING.COM WITH HTTPS.
THANKS FOR TRYING TO HELP, BUT I HAVE TO SAY GOOD NIGHT NOW.

**Comment by pstorch** [comment]
Related to #5800 I think.
If the default browser can't connect via https, then the SSL implementation on the phone is too old.

**Comment by rsudev** [comment]
It looks like Android 4.0.4 only supports TLS up to TLS 1.0 out of the box - and at least currently geocaching.com seems to require TLS 1.1 minimum (I checked that using the ssllabs service).
I assume that this is a recent change on there side.
I don't know whether we have a chance to mitigate the gap in Android from our side (like Chrome does)

**Comment by pstorch** [comment]
@rsudev you are right with Android 4.0.4. This topic sounded familiar to me, so I searched our code. We use the OkHttp library and about a year ago we put in extra code to enable Android 4.1.x devices to use newer TLS version. But this patch is not available below that API level.

**Comment by rsudev** [comment]
Reading again the stuff mentioned by SammysHP in #5800 I found the following link mentioned: https://blog.dev-area.net/2015/08/17/protect-your-android-app-against-ssl-exploits/
Looks like there might be a way to solve this with the help of GooglePlayServices...

**Comment by mucek4** [comment]
@rsudev What about devices that does not use Google services? Fix shall be easy if it really works?

**Comment by mucek4** [comment]
At home I have "Samsung Galaxy Core I8260" Android 4.1.2.
Did some quick test:
c:geo - Login OK
geocaching official app - Login fails (TLS 1.1 is supported from Androd 4.4)

So them might also have problems with older phones. 
Sadly I have no 4.0.x at home.

**Comment by Lineflyer** [comment]
We should make the first decision at first: Do we really want to fix it?

According to Google Play we have 3129 active installations for Android 4.0.x which is 0,4% of all active installations. There might be a number of side installs though. 
Is it worth the effort?

Maybe there is an external solution for the Android 4.0.x users (I read about an proxy app which handles TSL1.2 and forwards traffic).

**Comment by pstorch** [comment]
Facing these numbers I would like to vote for "Won't fix". 
As I mentioned above we already added some extra lines of code a year ago to support 4.1.x devices to support newer TLS standards.
If it might be solved with a single line of code like mentioned by @rsudev via link from @SammysHP, then maybe. But it only works for devices with google play services.
I definitely feel uncomfortable with a TLS proxy app.

**Comment by Lineflyer** [comment]
But if the workaround only applies for Google Play Services enabled devices, we might have an unclear situation whether or not people can still use the app.
It sounds like it makes sense to raise the minsdk anyway, but (if possible) still implement the mentioned workaround, so that user on 4.0.x with Play Services can side install the app.This looks to me as a better solution than offering the app to all potentially non working devices by default.
null

**Comment by mucek4** [comment]
If I read that right, It's not "workaround" for 4.0.x but an improvement for SSL security overall. I am not sure how this works without Play services, but it might be worth trying.
Second to that we can run the line of code if we detect Play services and skip it if not.

**Comment by Lineflyer** [comment]
But then the problem persists, that devices on 4.0.x without Play Services will not work (but still can install c:geo from Playstore).

**Comment by pstorch** [comment]
Something interesting from GS forum https://forums.geocaching.com/GC/index.php?/topic/348984-geocaching-no-longer-accessible-from-android-devices-with-version-lower-than-41/ and blog https://www.geocaching.com/blog/2018/07/security-update-required-for-android-devices/.

> The Payment Card Industry (PCI) Data Security Standard is a security standard for organizations that handle credit cards from major credit card providers. In order to remain compliant with these security standards, we had to stop supporting an older security protocol for interacting with our servers. In this case, we removed support for a standard called TLS 1.0. We stopped support on June 30th which can affect Android devices that run an OS of 5.1 or lower.

**Comment by Lineflyer** [comment]
Added FAQ entry http://www.cgeo.org/faq#7078

**Comment by mucek4** [comment]
@Lineflyer technically TLS 1.1 is natively supported from 4.4.x on, however we ( https://github.com/cgeo/cgeo/issues/7078#issuecomment-403243393 ) already have workaround for android 4.1 - 4.3

**Comment by Lineflyer** [comment]
I know, but I wanted to keep it easy saying.

**Comment by Lomanic** [comment]
I can see different ways to circumvent the obsolete system TLS stack, but they all would add complexity (of varying degree) unfortunately:

- Could creating a TLSConnection Factory with [the following](https://stackoverflow.com/a/33495988) work with https://rtyley.github.io/spongycastle/ on 4.0.4?
- Another idea would be to call an external program with its own TLS stack (static curl, or a Go program because of its easy cross-compilation and static builds out-of-the-box) with `Runtime.exec`, passing parameters to stdin and getting results (page source/files) from stdout.
- Another path could be to statically link cgeo to libcurl/OpenSSL via the Android NDK (https://github.com/bertrandmartel/ndk-curl-sample ?) and/or using JNI. That's how Firefox or Chrome can still connect to gc.com on 4.0.4 probably.

But of course as already said and in front of all this added complexity,  the easiest is to simply drop support for these old versions of Android, which is quite unfortunate for the owners of these devices.

**Comment by mucek4** [comment]
I have tried that 1 line "fix" mentioned above with google play services. It does not work. 
The code does some initialization, but SSL handshake still fails.

**Comment by Lineflyer** [comment]
If we do not find a quick and usable workaround, I suggest to use the next upcoming bugfix release to change the existing closeout notification system on the screen accordingly. So users will be notified, that their device will be/is no longer supported. In the version after that we can then raise 'minsdk'.

**Comment by Lineflyer** [comment]
I tried to take a look of how to adapt the ``status_closeout_warning`` to Android 4.1 accordingly. However it seems (with my basic knowlege) that only the strings exist but the logic in the code is (no longer) existing.
Can someone doublecheck? The goal would be, to show the closeout warning for all releases below 4.1 (and update the string accordingly).
